File: griffin/storage/graph.py
from __future__ import annotations

from pathlib import Path
from typing import Any, Dict, List, Optional
from .db import open_readonly

class GraphStore:
    def __init__(self, db_path: Optional[Path] = None):
        self.db_path = db_path

    def get_node(self, node_id: str) -> Optional[Dict[str, Any]]:
        conn = open_readonly(self.db_path)
        try:
            cursor = conn.execute(
                "SELECT id, kind, subtype, label, recorded_at, accession, authority, origin, review_state "
                "FROM v_node WHERE id = ?",
                (node_id,),
            )
            row = cursor.fetchone()
            if not row:
                return None
            return {
                "id": row[0],
                "kind": row[1],
                "subtype": row[2],
                "label": row[3],
                "recorded_at": row[4],
                "accession": row[5],
                "authority": row[6],
                "origin": row[7],
                "review_state": row[8],
            }
        finally:
            conn.close()

    def lineage(self, node_id: str, max_depth: int = 6) -> Dict[str, List[Dict[str, Any]]]:
        conn = open_readonly(self.db_path)
        try:
            sql = """
                WITH RECURSIVE anc(id, depth) AS (
                    SELECT ?, 0
                    UNION ALL
                    SELECT e.to_id, a.depth + 1
                    FROM v_edge_active e
                    JOIN anc a ON e.from_id = a.id
                    WHERE a.depth < ?
                )
                SELECT DISTINCT n.id, n.kind, n.subtype, n.label, n.recorded_at, n.origin, n.review_state
                FROM v_node n JOIN anc a ON n.id = a.id
            """
            cursor = conn.execute(sql, (node_id, max_depth))
            nodes = [
                {
                    "id": r[0],
                    "kind": r[1],
                    "subtype": r[2],
                    "label": r[3],
                    "recorded_at": r[4],
                    "origin": r[5],
                    "review_state": r[6],
                }
                for r in cursor.fetchall()
            ]
            
            node_ids = tuple(n["id"] for n in nodes)
            if not node_ids:
                return {"nodes": [], "edges": []}
            
            placeholders = ",".join("?" * len(node_ids))
            edges_cursor = conn.execute(
                f"SELECT id, from_id, to_id, relation, origin, confidence, created_at "
                f"FROM v_edge_active WHERE from_id IN ({placeholders}) AND to_id IN ({placeholders})",
                node_ids + node_ids,
            )
            edges = [
                {
                    "id": r[0],
                    "from_id": r[1],
                    "to_id": r[2],
                    "relation": r[3],
                    "origin": r[4],
                    "confidence": r[5],
                    "created_at": r[6],
                }
                for r in edges_cursor.fetchall()
            ]
            return {"nodes": nodes, "edges": edges}
        finally:
            conn.close()

File: griffin-next/backend/cli/src/server/routes/graph.ts
import { Hono } from "hono"
import { describeRoute, resolver, validator } from "hono-openapi"
import z from "zod"
import { DatabaseClient } from "@/storage/db/client"
import { GraphQuery } from "@/storage/db/graph/query"
import { lazy } from "@/util/lazy"

const SubgraphSchema = z.object({
  nodes: z.array(
    z.object({
      id: z.string(),
      kind: z.string(),
      subtype: z.string().nullable().optional(),
      label: z.string(),
      recorded_at: z.number(),
      origin: z.string(),
      review_state: z.string(),
      accession: z.string().nullable().optional(),
      authority: z.string().nullable().optional(),
    }),
  ),
  edges: z.array(
    z.object({
      id: z.number(),
      from_id: z.string(),
      to_id: z.string(),
      relation: z.string(),
      origin: z.string(),
      confidence: z.number().nullable().optional(),
      created_at: z.number(),
    }),
  ),
})

export const GraphRoutes = lazy(() =>
  new Hono()
    .get(
      "/lineage/:id",
      describeRoute({
        summary: "Get lineage subgraph for a node",
        description: "Returns recursive ancestry or descendants subgraph bounded by max depth.",
        operationId: "graph.lineage",
        responses: {
          200: {
            description: "Subgraph",
            content: { "application/json": { schema: resolver(SubgraphSchema) } },
          },
        },
      }),
      validator("param", z.object({ id: z.string().min(1) })),
      validator("query", z.object({ direction: z.enum(["ancestors", "descendants", "both"]).optional(), depth: z.string().optional() })),
      async (c) => {
        const id = c.req.valid("param").id
        const direction = c.req.valid("query").direction ?? "ancestors"
        const maxDepth = parseInt(c.req.valid("query").depth ?? "6", 10)
        const handle = DatabaseClient.reader()
        const res = GraphQuery.lineage(handle, { nodeId: id, direction, maxDepth })
        return c.json(res)
      },
    )
    .get(
      "/neighbors/:id",
      describeRoute({
        summary: "Get 1-hop neighbors for a node",
        description: "Returns 1-hop connected nodes and edges bounded by relation or kind.",
        operationId: "graph.neighbors",
        responses: {
          200: {
            description: "Subgraph",
            content: { "application/json": { schema: resolver(SubgraphSchema) } },
          },
        },
      }),
      validator("param", z.object({ id: z.string().min(1) })),
      validator("query", z.object({ relation: z.string().optional(), kind: z.string().optional() })),
      async (c) => {
        const id = c.req.valid("param").id
        const { relation, kind } = c.req.valid("query")
        const handle = DatabaseClient.reader()
        const res = GraphQuery.neighbors(handle, id, relation, kind)
        return c.json(res)
      },
    )
    .get(
      "/search",
      describeRoute({
        summary: "Search nodes",
        description: "Search nodes by keyword, accession, or label.",
        operationId: "graph.search",
        responses: {
          200: {
            description: "Matched nodes",
            content: { "application/json": { schema: resolver(z.object({ nodes: z.array(z.any()) })) } },
          },
        },
      }),
      validator("query", z.object({ q: z.string().min(1) })),
      async (c) => {
        const q = c.req.valid("query").q
        const handle = DatabaseClient.reader()
        const nodes = GraphQuery.search(handle, q)
        return c.json({ nodes })
      },
    ),
)
